android/src/main/java/com/swmansion/reanimated/layoutReanimation/SharedTransitionManager.java

```diff
@@ -0,0 +1,287 @@
+import com.facebook.react.views.view.ReactViewGroup;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+public class SharedTransitionManager {
+  private final AnimationsManager mAnimationsManager;
+  private final List<View> mAddedSharedViews = new ArrayList<>();
+  private final Map<Integer, View> mSharedTransitionParent = new HashMap<>();
+  private final Map<Integer, Integer> mSharedTransitionInParentIndex = new HashMap<>();
+  private boolean mIsSharedTransitionActive;
+  private final Map<Integer, Snapshot> mSnapshotRegistry = new HashMap<>();
+  private final List<View> mCurrentSharedTransitionViews = new ArrayList<>();
+  private View mTransitionContainer;
+  private final List<View> mRemovedSharedViews = new ArrayList<>();
+
+  public SharedTransitionManager(AnimationsManager animationsManager) {
+    mAnimationsManager = animationsManager;
+  }
+
+  protected void notifyAboutNewView(View view) {
+    mAddedSharedViews.add(view);
+  }
+
+  protected void notifyAboutRemovedView(View view) {
+    mRemovedSharedViews.add(view);
+  }
+
+  protected List<View> getCurrentSharedTransitionViews() {
+    return mCurrentSharedTransitionViews;
+  }
+
+  protected void viewsDidLayout() {
+    setupStaredTransitionForViews(mAddedSharedViews, true);
+    mAddedSharedViews.clear();
+  }
+
+  protected void viewsDidRemoved() {
+    setupStaredTransitionForViews(mRemovedSharedViews, false);
+    for (View removedSharedView : mRemovedSharedViews) {
+      visitTree(removedSharedView.getId(), true);
+    }
+    mRemovedSharedViews.clear();
+  }
+
+  private void setupStaredTransitionForViews(List<View> sharedViews, boolean withNewElements) {
+    if (sharedViews.isEmpty()) {
+      return;
+    }
+    sortViewsByTags(sharedViews);
+    List<SharedElement> sharedElements =
+        getSharedElementForCurrentTransition(sharedViews, withNewElements);
+    if (sharedElements.isEmpty()) {
+      return;
+    }
+    setupTransitionContainer();
+    reparentSharedViewsForCurrentTransition(sharedElements);
+    startSharedTransition(sharedElements);
+  }
+
+  private void sortViewsByTags(List<View> views) {
+    Collections.sort(views, (v1, v2) -> Integer.compare(v2.getId(), v1.getId()));
+  }
+
+  private List<SharedElement> getSharedElementForCurrentTransition(
+      List<View> sharedViews, boolean withNewElements) {
+    Set<Integer> viewTags = new HashSet<>();
+    if (!withNewElements) {
+      for (View view : sharedViews) {
+        viewTags.add(view.getId());
+      }
+    }
+    List<SharedElement> sharedElements = new ArrayList<>();
+    ReanimatedNativeHierarchyManager reanimatedNativeHierarchyManager =
+        mAnimationsManager.getReanimatedNativeHierarchyManager();
+    for (View sharedView : sharedViews) {
+      int targetViewTag =
+          mAnimationsManager
+              .getNativeMethodsHolder()
```

Comment:
This getter is called a lot. Maybe it's worth storing the `NativeMethodsHolder` in a field?